src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -3962,25 +3985,42 @@ LogicalResult ONNXGreaterOrEqualOp::inferShapes(
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// ONNXHardmaxOp
+//===----------------------------------------------------------------------===//
+
 LogicalResult ONNXHardmaxOp::verify() {
-  ONNXHardmaxOpAdaptor hmOp = ONNXHardmaxOpAdaptor(*this);
-  auto input = hmOp.input();
-  int64_t axis = this->axis();
+  ONNXHardmaxOpAdaptor operandAdaptor(*this);
+  auto input = operandAdaptor.input();
+  if (!hasShapeAndRank(input))
+    // Won't be able to do any checking at this stage.
+    return success();
 
-  // Verify that axis must be in range [-r, r - 1], where r is the rank of
-  // input.
-  if (hasShapeAndRank(input)) {
-    int64_t rank = input.getType().cast<ShapedType>().getRank();
-    if (axis < -rank || axis > rank - 1)
-      return emitOpError("axis value is out of range");
-  }
+  // axis attribute must be in the range [-r,r-1], where r = rank(input).
+  int64_t axisValue = axis();
+  int64_t inputRank = input.getType().cast<ShapedType>().getRank();
+  if (axisValue < -inputRank || axisValue >= inputRank)
+    return onnx_mlir::Diagnostic::attributeOutOfRange(*this->getOperation(),
+        "axis", axisValue,
+        onnx_mlir::Diagnostic::Range<int64_t>(-inputRank, inputRank - 1));
 
   return success();
 }
 
 LogicalResult ONNXHardmaxOp::inferShapes(
     std::function<void(mlir::Region &)> doShapeInference) {
-  getResult().setType(getOperand().getType());
+  auto inputType = input().getType().cast<ShapedType>();
```

Comment:
During verification we can diagnose that the `rank` is out of range only if we know the range of the input tensor(s). 
In some cases the rank is known only after shape inference. For example (see issue https://github.com/onnx/onnx-mlir/issues/1280):

```
// -----// IR Dump Before (anonymous namespace)::ShapeInferencePass //----- //
 func @main_graph(.....) {
 ...
     %76 = "onnx.Concat"(%66, %75, %22, %24, %39, %41, %15, %17, %31, %33, %6, %57, %55, %53, %51, %49, %47) {axis = 33554432 : si64, onnx_node_name = "StatefulPartitionedCall/model_4/tf.concat_1/concat"} : (tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>) -> tensor<*xf32>
```
Although the `axis` value looks too large, we cannot diagnose it because the rank of the inputs is not yet known. 

```
// -----// IR Dump After (anonymous namespace)::ShapeInferencePass Failed //----- //
"builtin.module"() ({
  "builtin.func"() ({
  ...
      %76 = "onnx.Concat"(%66, %75, %22, %24, %39, %41, %15, %17, %31, %33, %6, %57, %55, %53, %51, %49, %47) {axis = 33554432 : si64, onnx_node_name = "StatefulPartitionedCall/model_4/tf.concat_1/concat"} : (tensor<?x7x1xf32>, tensor<?x7x1xf32>, tensor<?x7x1xf32>, tensor<?x7x1xf32>, tensor<?x7x1xf32>, tensor<?x7x1xf32>, tensor<?x7x1xf32>, tensor<?x7x1xf32>, tensor<?x7x1xf32>, tensor<?x7x1xf32>, tensor<?x7x3xf32>, tensor<?x7x6xf32>, tensor<?x7x7xf32>, tensor<?x7x2xf32>, tensor<?x7x7xf32>, tensor<?x7x3xf32>, tensor<?x7x1xf32>) -> tensor<*xf32>
```
Now is clear that the `axis` value (33554432) is out of the expected range [-3,2].

The error we emit at this point is:
```
error: onnx.Concat 'axis' value is 33554432, accepted range is [-3, 2]  
```

Reply:
I maintain that this code is broken. First of all, we could have inputs whose ranks are unknown. The approach that @chentong319 recommends is to return success, as after the shape inference stabilizes to a fix point, there is a pass that checks that all tensors are ranked. If that does not work, why is that the case?

As is, the code asserts that there is a rank.
Second, if it is ranked, the when that operation was created, its verifier would have worked.

Wasn't the issue to begin with that we had a rank of >1000 and we wanted to emit a clean error when seen these cases?

Reply:
@AlexandreEichenberger PR https://github.com/onnx/onnx-mlir/pull/1358 is an attempt to rectify the problem. Please take a look. We can return to this PR after that one lands.

Reply:
PR https://github.com/onnx/onnx-mlir/pull/1358 has landed, we can continue this review @AlexandreEichenberger